### PR TITLE
plutus-core: remove the value restriction in favour of big-lambdas computing

### PR DESCRIPTION
**Pull request by michaelpj**
(This is a preview, assuming we decide to go with this approach.)

This removes the value restriction from Plutus Core. This requires us to
decide when the bodies of type abstractions reduce. We choose to say that
they work like the bodies of lambda abstractions: they reduce after the
substitution created by the abstraction is applied.

As it happens, this requires few changes:
- The reduction rules for type abstractions already reveal the body
after the substitution and not before - since there is no reduction
frame for type abstractions, this has the behaviour we want.
- The specification of the CK machine is changed to compute the body of
the abstraction after the instantiation.
    - As it turns out, the implementations of all our machines
already do this. Currently this is redundant since it will already be a
value, but it means we don't need to change anything.
- We don't need to say anything about erasure, since we have neither a
specification of it nor an implementation at the moment. There will need
to be change in the metatheory, however.

### CODE REVIEW

File: plutus-core-spec/plutus-core-specification.tex [hunk 1]
```diff
@@ -332,10 +332,8 @@ \subsection{Notes on the grammar}
 \item $\con{cn}$ represents a constant belonging to some built-in
   type: for example \texttt{(con 4)} is an integer with the value 127.
   See Section~\ref{sec:builtins} for more on this.
-\item $\abs{\alpha}{K}{V}$ represents a polymorphic value abstracted
-  over a type; this would often be donated by $\Lambda\alpha{::}K.V$.
-  Note that we require the body of the abstraction to be a
-  \textit{value}, not a term: see the note on the value restriction below.
+\item $\abs{\alpha}{K}{M}$ represents a polymorphic value abstracted
+  over a type; this would often be donated by $\Lambda\alpha{::}K.M$.
```
Comment (kwxm) [inline plutus-core-spec/plutus-core-specification.tex]: Donated?

**Comment by michaelpj** [comment]
> I think it's a weird semantics. So we have
> const {integer} {forall a. a} 42 (/\a -> error <a>)
> evaluating to Failure, right? Reducing /\a -> error <a> even though the type of the argument clearly says that the big lambda shouldn't be stripped off, feels really weird to me.

No, this evaluates to `42`. This is the semantics where we evaluate the body of a big lambda after it is instantiated. Which is the one that both you and I like ;)

I'll add the test case just to prove it.

**Comment by effectfully** [comment]
> No, this evaluates to 42. This is the semantics where we evaluate the body of a big lambda after it is instantiated. Which is the one that both you and I like ;)

Ah, stupid me then.

**Comment by kwxm** [comment]
There is a price to be paid here if we're thinking about erasing things.  If you look at https://github.com/input-output-hk/plutus/blob/master/docs/plutus-core/Merklisation/AST-analysis.md you'll see that in our validators, 18-22% of the `Term` nodes are `TyAbs` or `TyInst`.  If we have the value restriction or allow computation under `Λ` then we can happily discard all of these without changing the semantics of a program. With the proposal here for removing  the value restriction you have to insert something in the AST to mimic the computational effects of the type-level operations. I ran some experiments to see what happened.  I tried two different things: introducing new `Delay` and `Force` terms into the AST type, and replacing `Λt::k.e` with `λu.e` and `{e t}` with `[e u]`, for a suitable unit value `u`.  The effects of these are shown in the table below (first column of results: size of unmodified AST; second column: size after erasing all the usual stuff, assuming we have the value restriction; third column: size after replacing type-level operations with `Force` and `Delay` and erasing everything else; final column: size after replacing type-level operations with term-level ones involving units).

Conclusion: using `Force/Delay` increases the size of compressed minimised validators by about 10% and using type-level operations by about 15% (and the uncompressed size much more).  We can probably live with this, and `Force/Delay` looks more economical.

| Contract | Compression | Typed | Minimised, value restriction | Miminimsed, Force/Delay | Minimised, lambda/apply |
| :---: | ---: | ---: | ---: | ---: | ---: | 
| Crowdfunding | Uncompressed | 110413 | 6593 (6.0%) | 7471 (6.8%) | 10056 (9.1%) |        
|              | Compressed   | 21440  | 1541 (1.4%) | 1698 (1.5%) | 1769 (1.6%) |         
| |
| |
| Currrency | Uncompressed | 159160 | 10655 (6.7%)   | 12135 (7.6%) | 16942 (10.6%) |       
|           | Compressed   | 29004  | 3010 (1.9%)    | 3267 (2.1%)  | 3473 (2.2%) |         
| |
| |
| Escrow | Uncompressed | 117438 | 7653 (6.5%) 	     | 8616 (7.3%) | 11571 (9.9%) |        
|        | Compressed | 22064 | 1882 (1.6%)  	     | 2054 (1.7%) | 2123 (1.8%) |         
| |
| |
| Future | Uncompressed | 206137 | 13914 (6.7%)      | 15692 (7.6%) | 21400 (10.4%) |       
|        | Compressed | 40134 | 3394 (1.6%) 	     | 3732 (1.8%)  | 3911 (1.9%) |         
| |
| |
| Game | Uncompressed | 98970 | 6635 (6.7%) 	     | 7508 (7.6%)  | 10382 (10.5%) |       
|      | Compressed | 16898 | 2148 (2.2%)            | 2369 (2.4%)  | 2516 (2.5%) |         
| |
| |
| GameStateMachine | Uncompressed | 117053 | 8997 (7.7%) | 10196 (8.7%) | 14092 (12.0%) |       
|                  |   Compressed | 22322 | 1980 (1.7%)  | 2182 (1.9%)  | 2276 (1.9%) |         
| |
| |
| MultiSig | Uncompressed | 111990 | 8149 (7.3%) | 9202 (8.2%) | 12676 (11.3%) |       
|          |   Compressed | 19771 | 2552 (2.3%) | 2777 (2.5%)  | 2942 (2.6%) |         
| |
| |
| MultiSigStateMachine | Uncompressed | 196628 | 14956 (7.6%) | 16843 (8.6%) | 22960 (11.7%) |       
|                      |   Compressed | 37535 | 3719 (1.9%)   | 4066 (2.1%)  | 4286 (2.2%) |         
| |
| |
| PubKey | Uncompressed | 106535 | 7579 (7.1%) | 8560 (8.0%) | 11806 (11.1%) |       
|        |   Compressed | 18483  | 2409 (2.3%) | 2627 (2.5%) | 2780 (2.6%) |         
| |
| |
| Swap | Uncompressed | 154724 | 12077 (7.8%) | 13507 (8.7%) | 18081 (11.7%) |       
|      |   Compressed | 28243 | 3756 (2.4%)   | 4051 (2.6%)  | 4310 (2.8%) |         
| |
| |
| TokenAccount | Uncompressed | 78030 | 3184 (4.1%) | 3744 (4.8%) | 5270 (6.8%) |         
|              |   Compressed | 15765 | 659 (0.8%)  | 758 (1.0%)  | 793 (1.0%) |          
| |
| |
| Vesting | Uncompressed | 119514 | 6967 (5.8%) | 7920 (6.6%) | 10736 (9.0%) |        
|         |   Compressed | 23348  | 1614 (1.4%) | 1775 (1.5%) | 1875 (1.6%) |

**Comment by michaelpj** [comment]
I think for the size comparison it's critical to let-bind (well, lambda-bind) the unit argument. If we do that, then we should be replacing a type node with a variable node, which should be smaller if anything. Otherwise we're inserting a whole unit term definition, which is probably often bigger.

The runtime changes are worth thinking about, though.

**Comment by kwxm** [comment]
> I think for the size comparison it's critical to let-bind (well, lambda-bind) the unit argument. 

You're right, of course.  I reran it with that strategy (but not actually wrapping the body of the program in a lambda binding the unit value, which would add a few more nodes/bytes), and the results are in between the last two columns.  The compressed sizes lie 20-50% of the way from the second last column to the last one, if you see what I mean.

**Comment by michaelpj** [comment]
Okay, I think everyone agrees we want to do this from a semantic point of view. It's good to note that this will make the erased size bigger, but I think we can deal with that later if we need to.

**Pull request merged**